### PR TITLE
Remove hardcoded OID in GPORCA and add fallback inside GPDB translator

### PR DESCRIPTION
**Pull request by vraghavan78**
The OID changed in GPDB 6. Opened a PR inside GPORCA to remove all hardcoded stuff.

### CODE REVIEW

**Comment by vraghavan78** [comment]
Opened an issue https://github.com/greenplum-db/gpdb/issues/4085

**Comment by vraghavan78** [comment]
Related GPORCA changes are
https://github.com/greenplum-db/gporca/pull/273

File: src/backend/gpopt/translate/CTranslatorScalarToDXL.cpp [hunk 1]
```diff
@@ -1564,6 +1564,19 @@ CTranslatorScalarToDXL::PdxlnScWindowFunc
 
 	GPOS_ASSERT(EdxlwinstageSentinel != edxlwinstage && "Invalid window stage");
 
+	if (3103 == pwindowfunc->winfnoid)
```
Comment (hlinnaka) [inline src/backend/gpopt/translate/CTranslatorScalarToDXL.cpp]: I don't understand this. Is the intention to force fallback for any queries that use these three functions? Why, ORCA used to handle them just fine, no? Are ROW_NUMBER() and DENSE_RANK() still OK?

Should we rather remove the mapping for these functions from CTranslatorRelcacheToDXL.cpp, or throw an error there directly?

Comment (vraghavan78) [reply src/backend/gpopt/translate/CTranslatorScalarToDXL.cpp]: We never supported them please see the orca changes. The fallback was inside orca with hard coded oids

Comment (vraghavan78) [reply src/backend/gpopt/translate/CTranslatorScalarToDXL.cpp]: Row number and dense rank are ok

**Comment by hlinnaka** [comment]
On 08/12/17 10:49, Venkatesh Raghavan wrote:
> vraghavan78 commented on this pull request.
>> @@ -1564,6 +1564,19 @@ CTranslatorScalarToDXL::PdxlnScWindowFunc
>   
>   	GPOS_ASSERT(EdxlwinstageSentinel != edxlwinstage && "Invalid window stage");
>   
> +	if (3103 == pwindowfunc->winfnoid)
> 
> We never supported them please see the orca changes. The fallback was inside orca with hard coded oids

Oh, ok. How about catching these in 
CTranslatorRelcacheToDXL::PimdobjGPDB and 
CTranslatorRelcacheToDXL::Pmdfunc already, then?

Does ORCA support arbitrary user-defined ordered-set aggregates, or 
should we fall back on those as well?

Actually, ORCA currently seems to fall back on any ordered-set 
aggregates (i.e. WITHIN GROUP (...)) Is that right? If so, can we just 
remove the mapping for those three functions, and have ORCA fall back 
like it does for any ordered-set aggregates?

- Heikki

**Comment by vraghavan78** [comment]
Yes we fallback for ordered aggregates in CTranslstorScalarToDXL

‘’’	if (paggref->aggorder != NIL)
	{
		GPOS_RAISE
		(
				gpdxl::ExmaDXL,
				gpdxl::ExmiPlStmt2DXLConversion,
				GPOS_WSZ_LIT("Ordered aggregates")
		);
	}
‘’’

So what is the property we should check for the fallback?

**Comment by vraghavan78** [comment]
Just noticed the error type needs to change will fix it

**Comment by vraghavan78** [comment]
@hlinnaka consider the following query:

```sql
explain SELECT PERCENT_RANK() OVER(win2)                                                                                                                                                       FROM (SELECT cf_olap_windowerr_sale_ord.* FROM cf_olap_windowerr_sale_ord) cf_olap_windowerr_sale                                                                                                           WINDOW win2 as (order by cf_olap_windowerr_sale.vn desc);
```
The window function has the following information in the parse tree.

```
{WINDOWFUNC
         :winfnoid 3103
         :wintype 701
         :args <>
         :aggfilter <>
         :winref 1
         :winstar false
         :winagg false
         :windistinct false
         :winindex 0
         :winstage 0
         :location 15
         }
```

I am not sure beside Oid what I can use to generalize not supporting this feature.

For  arbitrary user-defined ordered-set aggregates or regular ordered aggregates we have a list of order clause that can be used to fall back.

```
typedef struct Aggref
{
	Expr		xpr;
	Oid			aggfnoid;		/* pg_proc Oid of the aggregate */
	Oid			aggtype;		/* type Oid of result of the aggregate */
	List	   *aggdirectargs;	/* direct arguments, if an ordered-set agg */
	List	   *args;			/* aggregated arguments and sort expressions */
	List	   *aggorder;		/* ORDER BY (list of SortGroupClause) */
	List	   *aggdistinct;	/* DISTINCT (list of SortGroupClause) */
	Expr	   *aggfilter;		/* FILTER expression, if any */
	bool		aggstar;		/* TRUE if argument list was really '*' */
	bool		aggvariadic;	/* true if variadic arguments have been
								 * combined into an array last argument */
	char		aggkind;		/* aggregate kind (see pg_aggregate.h) */
	Index		agglevelsup;	/* > 0 if agg belongs to outer query */
	AggStage	aggstage;		/* MPP: 2-stage? If so, which stage */
	int			location;		/* token location, or -1 if unknown */
} Aggref;
```

**Comment by hlinnaka** [comment]
On 08/12/17 19:20, Venkatesh Raghavan wrote:
> @hlinnaka consider the following query:
> 
> ```sql
> explain SELECT PERCENT_RANK() OVER(win2)                                                                                                                                                       FROM (SELECT cf_olap_windowerr_sale_ord.* FROM cf_olap_windowerr_sale_ord) cf_olap_windowerr_sale                                                                                                           WINDOW win2 as (order by cf_olap_windowerr_sale.vn desc);
> ```
> The window function has the following information in the parse tree.
> 
> ```
> {WINDOWFUNC
>           :winfnoid 3103
>           :wintype 701
>           :args <>
>           :aggfilter <>
>           :winref 1
>           :winstar false
>           :winagg false
>           :windistinct false
>           :winindex 0
>           :winstage 0
>           :location 15
>           }
> ```
> 
> I am not sure beside Oid what I can use to generalize not supporting this feature.

I still don't understand. What exactly is the feature that ORCA doesn't 
support?

> For  arbitrary user-defined ordered-set aggregates or regular ordered aggregates we have a list of order clause that can be used to fall back.

It's entirely possible to create a user-defined function that's 
identical to PERCENT_RANK():

create function my_percent_rank() returns float8 window immutable 
language internal as 'window_percent_rank';

- Heikki

**Comment by vraghavan78** [comment]
@hlinnaka ur correct. I was merely doing a lift a shift from old logic to accommodate the merge issues. We should re-visit this. closing this PR

**Pull request closed without merging**